### PR TITLE
global: serializers refactoring

### PR DESCRIPTION
**Pull request by krzysztof**
* Refactoring of serializers.

### CODE REVIEW

**Comment by krzysztof** [comment]
Addresses #39

File: invenio_records_rest/serializers/__init__.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Record serialization."""
+
+from __future__ import absolute_import, print_function
+
+from dojson.contrib.to_marc21 import to_marc21
+
+from .json import JSONSerializer
+from .marcxml import MARCXMLSerializer
```
Comment (lnielsen) [inline invenio_records_rest/serializers/__init__.py]: This should probably go to invenio-marc21.

File: invenio_records_rest/serializers/__init__.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+from .response import record_responsify, search_responsify
+from .schemas.json import RecordSchemaJSONV1
+from .schemas.marcxml import RecordSchemaMARC
```
Comment (lnielsen) [inline invenio_records_rest/serializers/__init__.py]: This should probably go to invenio-marc21.

File: invenio_records_rest/serializers/__init__.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+json_v1 = JSONSerializer(RecordSchemaJSONV1)
+marcxml_v1 = MARCXMLSerializer(to_marc21, schema_class=RecordSchemaMARC)
```
Comment (lnielsen) [inline invenio_records_rest/serializers/__init__.py]: invenio-marc21 (ditto for the other marcxml stuff)

File: invenio_records_rest/serializers/dojson.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
```
Comment (lnielsen) [inline invenio_records_rest/serializers/dojson.py]: Should go to invenio-marc21.

File: invenio_records_rest/serializers/response.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    :param serializer: Serializer instance.
+    :param mimetype: MIME type of response.
+    """
+    def view(pid, record, code=200, headers=None):
+        links = {}
+        response = current_app.response_class(
+            serializer.serialize(pid, record, links=links),
+            mimetype=mimetype)
+        response.status_code = code
+        # response.headers['location'] = links['self']
```
Comment (lnielsen) [inline invenio_records_rest/serializers/response.py]: You can take this line out completely. It should only be used in 201/3xx responses.

File: invenio_records_rest/serializers/schemas/json.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    grants = fields.List(fields.Str)
+
+    communities = fields.List(fields.Str)
+
+    references = fields.List(fields.Str)
+    journal = fields.Nested(JournalSchemaV1)
+    conference = fields.Nested(ConferenceSchemaV1)
+
+
+class RecordSchemaJSONV1(Schema):
```
Comment (lnielsen) [inline invenio_records_rest/serializers/schemas/json.py]: This file is naturally Zenodo specific, but you can basically trim it down to this class - just change ``metadata = fields.Raw()``

File: invenio_records_rest/serializers/schemas/marcxml.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Zenodo; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""MARCXML translation index."""
```
Comment (lnielsen) [inline invenio_records_rest/serializers/schemas/marcxml.py]: This one is also Zenodo specific, and I don't think we need to move it. In the MARC21 package you can assume that the JSON structure you have contains the right data (i.e. no need for date time formatting on the fly).

File: tests/test_invenio_records_rest.py [hunk 8]
```diff
@@ -265,6 +265,7 @@ def test_valid_get(app):
             res = client.get(url_for('invenio_records_rest.recid_item',
                                      pid_value=pid.pid_value),
                              headers=headers)
+            import ipdb; ipdb.set_trace()
```
Comment (lnielsen) [inline tests/test_invenio_records_rest.py]: :-)

File: invenio_records_rest/serializers/__init__.py (hunk 1 above)
Comment (krzysztof) [inline invenio_records_rest/serializers/__init__.py]: Should then ``invenio-marc21`` depend on ``invenio-records-rest`` ?

Comment (lnielsen) [inline invenio_records_rest/serializers/__init__.py]: It doesn't need a hard dependency. Basically, marc21 can just have an example RECORDS_REST_ENDPOINTS configuration that uses the marc serializer

File: setup.py [hunk 9]
```diff
@@ -63,12 +63,15 @@
 ]
 
 install_requires = [
+    'dojson>=1.0.0',
```
Comment (lnielsen) [inline setup.py]: As ``DoJSONSerializer`` goes to invenio-marc21, then this one can probably be taken out

File: tests/test_serializer_base.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Zenodo.
```
Comment (lnielsen) [inline tests/test_serializer_base.py]: Wrong license header.

File: tests/test_serializer_base.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Zenodo; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Zenodo serializer tests."""
```
Comment (lnielsen) [inline tests/test_serializer_base.py]: ``Serializer tests.``

File: tests/test_serializer_json.py [hunk 12]
```diff
@@ -0,0 +1,103 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Zenodo.
```
Comment (lnielsen) [inline tests/test_serializer_json.py]: Wrong license header.

File: tests/test_serializer_json.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Zenodo; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Zenodo serializer tests."""
```
Comment (lnielsen) [inline tests/test_serializer_json.py]: Ditto

File: tests/test_serializer_marshmallow.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Zenodo.
```
Comment (lnielsen) [inline tests/test_serializer_marshmallow.py]: Wrong license header.

File: tests/test_serializer_marshmallow.py [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Zenodo; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Zenodo serializer tests."""
```
Comment (lnielsen) [inline tests/test_serializer_marshmallow.py]: Ditto

File: tests/test_serializer_response.py [hunk 16]
```diff
@@ -0,0 +1,83 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Zenodo.
```
Comment (lnielsen) [inline tests/test_serializer_response.py]: Ditto

File: invenio_records_rest/serializers/__init__.py [hunk 17]
```diff
@@ -0,0 +1,35 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Zenodo.
```
Comment (lnielsen) [inline invenio_records_rest/serializers/__init__.py]: Wrong license header.

File: invenio_records_rest/serializers/base.py [hunk 18]
```diff
@@ -0,0 +1,67 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Zenodo.
```
Comment (lnielsen) [inline invenio_records_rest/serializers/base.py]: Wrong license header.

File: invenio_records_rest/serializers/json.py [hunk 19]
```diff
@@ -0,0 +1,86 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Zenodo.
```
Comment (lnielsen) [inline invenio_records_rest/serializers/json.py]: Wrong license header.

File: invenio_records_rest/serializers/marshmallow.py [hunk 20]
```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Zenodo.
```
Comment (lnielsen) [inline invenio_records_rest/serializers/marshmallow.py]: Wrong license header.

File: invenio_records_rest/serializers/response.py [hunk 21]
```diff
@@ -0,0 +1,70 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Zenodo.
```
Comment (lnielsen) [inline invenio_records_rest/serializers/response.py]: Wrong license header.

File: invenio_records_rest/serializers/schemas/__init__.py [hunk 22]
```diff
@@ -0,0 +1,27 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Zenodo.
```
Comment (lnielsen) [inline invenio_records_rest/serializers/schemas/__init__.py]: Wrong license header.

File: invenio_records_rest/serializers/schemas/json.py [hunk 23]
```diff
@@ -0,0 +1,39 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Zenodo.
```
Comment (lnielsen) [inline invenio_records_rest/serializers/schemas/json.py]: Wrong license header.

File: tests/conftest.py [hunk 24]
```diff
@@ -115,15 +114,19 @@ def app(request):
...
-    def finalize():
-        with app.app_context():
-            db.drop_all()
-            if app.config['SQLALCHEMY_DATABASE_URI'] != 'sqlite://':
-                drop_database(db.engine.url)
-            shutil.rmtree(instance_path)
+    with app.app_context():
+        # Yield app in request context
+        # with app.test_request_context():
+        with app.test_request_context():  # TODO: Move to app_request fixture
```
Comment (lnielsen) [inline tests/conftest.py]: TODO?

**Comment by lnielsen** [comment]
Looks good. Thanks a lot for also fixing the import sorting problems. Only problems as I can see:

- [ ] Move DoJSON serialiser to Invenio-MARC21.
- [ ] Fix license headers.

Can you also add a ``Co-authored-by: Lars Holm Nielsen <lars.holm.nielsen@cern.ch>`` to the bottom.

**Comment by kaplun** [comment]
cc: @jmartinm (since you were starting to play with this part)

**Comment by krzysztof** [comment]
Fixed license headers, DoJSON moved and the TODO removed.

**Comment by krzysztof** [comment]
ES timing out...

**Pull request merged**